src/main/java/org/json/JSONArray.java

```diff
@@ -224,6 +226,14 @@ public JSONArray(int initialCapacity) throws JSONException {
     	this.myArrayList = new ArrayList<Object>(initialCapacity);
     }
 
+    /**
+     * Stream support - requires java 8
+     * @return Stream of array elements
+     */
+    public Stream<Object> stream() {
+        return IntStream.range(0, length()).mapToObj(this::get);
```

Comment:
> I think this should not be `this::get` and instead should be `this::opt`. Stream should not throw exceptions, which `get` will do if a value is `null`.
> 
> We should also be sure to include at least one test case that has `null` in the array.

Seems to be working as is, but not sure why. Test cases are updated, let me know if this code or the tests should still be changed.

Reply:
I stepped through the code and that Parser is converting `null` to `JSONObject.NULL`, which is expected.

Let's add these 2 additional tests, which are very similar, but add different ways of composing the data instead of just using the parser. This second test shows the issue I was worried about:

```java
@Test
    public void testStream_ComposeFromArray() {
        @SuppressWarnings("boxing")
        Object[] expectedList = new Object[] { 1, 2, 3, true, false, null, "abc", "def", new JSONObject("{ \"a\": \"b\" }"), new JSONArray()};
        JSONArray jsonArray = new JSONArray(expectedList);
        
        List<Object> actualList = jsonArray.stream()
                .collect(Collectors.toList());
        System.out.println(actualList);
        assertEquals(actualList.size(), expectedList.length);
        for (int i = 0; i < actualList.size(); ++i) {
            Object actual = actualList.get(i);
            Object expected = expectedList[i];
            if (actual instanceof JSONObject) {
                assertTrue("Json Object test",((JSONObject) actual).similar(expected));
            } else if (actual instanceof JSONArray) {
                assertTrue("Json Array test", ((JSONArray) actual).similar(expected));
            } else if (JSONObject.NULL.equals(expected)) {
                assertTrue("Null test", JSONObject.NULL.equals(actual));
            } else {
                assertEquals("value test: "+expected, expected, actual);
            }
        }
    }
```

```java
    @Test
    public void testStream_ComposeFromCode() {
        JSONArray jsonArray = new JSONArray();
        jsonArray.put(1).put(2).put(3)
            .put(true).put(false)
            .put((Object)null)
            .put("abc").put("def")
            .put(new JSONObject("{ \"a\": \"b\" }"))
            .put(new JSONArray())
        ;
        @SuppressWarnings("boxing")
        Object[] expectedList = new Object[] { 1, 2, 3, true, false, null, "abc", "def", new JSONObject("{ \"a\": \"b\" }"), new JSONArray()};
        
        List<Object> actualList = jsonArray.stream()
                .collect(Collectors.toList());
        System.out.println(actualList);
        assertEquals(actualList.size(), expectedList.length);
        for (int i = 0; i < actualList.size(); ++i) {
            Object actual = actualList.get(i);
            Object expected = expectedList[i];
            if (actual instanceof JSONObject) {
                assertTrue("Json Object test",((JSONObject) actual).similar(expected));
            } else if (actual instanceof JSONArray) {
                assertTrue("Json Array test", ((JSONArray) actual).similar(expected));
            } else if (JSONObject.NULL.equals(expected)) {
                assertTrue("Null test", JSONObject.NULL.equals(actual));
            } else {
                assertEquals("value test: "+expected, expected, actual);
            }
        }
    }
```

Reply:
Added tests